### PR TITLE
Adding caller attributes

### PR DESCRIPTION
**Pull request by CoryCharlton**
## Description
- Added `CallerArgumentExpressionAttribute`.
- Added `CallerMemberNameAttribute`.

## Motivation and Context
Allows for cleaner code.

## How Has This Been Tested?
- Unit tests
- Production use in another project

## Types of changes

- [X] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [X] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [X] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [X] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [X] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [X] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [X] I have added new tests to cover my changes.

### CODE REVIEW

**Comment by CoryCharlton** [comment]
Regarding the motivation and context the first place I want to use `CallerArgumentExpressionAttribute` would be in the TestFramework methods like `Assert.IsFalse`. This would allow a default value to be set for the message parameter so the exact expression that failed would be known.

File: System.Runtime/CallerArgumentExpressionAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+// ReSharper disable once CheckNamespace
+namespace System.Runtime.CompilerServices
+{
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
```
Comment (josesimoes) [inline System.Runtime/CallerArgumentExpressionAttribute.cs]: Why not adding the Intellisense comments?

File: System.Runtime/CallerMemberNameAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+// ReSharper disable once CheckNamespace
+namespace System.Runtime.CompilerServices
+{
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
```
Comment (josesimoes) [inline System.Runtime/CallerMemberNameAttribute.cs]: Same here

File: System.Runtime.UnitTests/CallerArgumentExpressionAttributeTests.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+{
+    [TestClass]
+    public class CallerArgumentExpressionAttributeTests
+    {
+        public string Field = "Field value";
+
+        public string Property => "Property value";
+
+        [TestMethod]
+        public void CallerArgumentExpression_sets_ParameterName_from_field()
```
Comment (josesimoes) [inline System.Runtime.UnitTests/CallerArgumentExpressionAttributeTests.cs]: Project code style calls for camel case method name. I know most of the mscorlib unit tests are using this naming style, but we should default to the appropriate one when adding new tests. Moreover, this is a brand-new project, so let's have it by the book, OK? 😉

**Review by josesimoes — COMMENTED** [review]
Adding to the review comments, now that there are unit tests, please enable the unit test running on the azure pipeline. 
If needed check an usage example [here](https://github.com/nanoframework/nanoFramework.Azure.Devices/blob/cec42a0176d0bca24925a449e5ffb188d4e0ace0/azure-pipelines.yml#L58).

File: System.Runtime/CallerArgumentExpressionAttribute.cs (hunk 1 above)
Comment (CoryCharlton) [reply System.Runtime/CallerArgumentExpressionAttribute.cs]: I was just copy/pasting from the reference code. I updated it to have the documentation from the web.

**Comment by CoryCharlton** [comment]
> Adding to the review comments, now that there are unit tests, please enable the unit test running on the azure pipeline. If needed check an usage example [here](https://github.com/nanoframework/nanoFramework.Azure.Devices/blob/cec42a0176d0bca24925a449e5ffb188d4e0ace0/azure-pipelines.yml#L58).

Sounds good. Not sure how to test the pipeline change but I made it and the other requested changes.

**Comment by josesimoes** [comment]
> Sounds good. Not sure how to test the pipeline change but I made it and the other requested changes.

It will run as part of the pipeline. If any unit tests fails, the build will fail.
Currently something off with teh AZO trigger... I'll look into it asap.

File: System.Runtime.UnitTests/nano.runsettings [hunk 4]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RunSettings>
+   <!-- Configurations that affect the Test Framework -->
+   <RunConfiguration>
```
Comment (networkfusion) [inline System.Runtime.UnitTests/nano.runsettings:4]: Try adding `<MaxCpuCount>1</MaxCpuCount>`

File: System.Runtime.UnitTests/nano.runsettings [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+       <ResultsDirectory>.\TestResults</ResultsDirectory><!-- Path relative to solution directory -->
+       <TestSessionTimeout>120000</TestSessionTimeout><!-- Milliseconds -->
+       <TargetFrameworkVersion>net48</TargetFrameworkVersion>
+       <TargetPlatform>x64</TargetPlatform>
+   </RunConfiguration>
+   <nanoFrameworkAdapter>
+       <Logging>None</Logging> <!--Set to the desired level of logging for Unit Test execution. Possible values are: None, Detailed, Verbose, Error. -->
+       <IsRealHardware>False</IsRealHardware><!--Set to true to run tests on real hardware. -->
+       <RealHardwarePort>COM3</RealHardwarePort><!--Specify the COM port to use to connect to a nanoDevice. If none is specified, a device detection is performed and the 1st available one will be used. -->
```
Comment (networkfusion) [inline System.Runtime.UnitTests/nano.runsettings:13]: Try removing these.

**Comment by networkfusion** [comment]
feel free to merge in `test-update-pr2` branch, which should fix all issues with the build..

**Comment by CoryCharlton** [comment]
> feel free to merge in `test-update-pr2` branch, which should fix all issues with the build..

I merged that branch but there still appear to be issues. Looks like the pipeline build isn't setup all the way (ie: the first error seems to be missing the github token): https://dev.azure.com/nanoframework/System.Runtime/_build/results?buildId=47240&view=results

**Comment by josesimoes** [comment]
Merging despite AZDO reporting failed run. Build was fine, new Unit Tests run successfully. 
This is due to a permission issue caused by running a PR from a forked repo.

**Pull request merged**